Update jobs metrics for pts & cdc gc_expire option

kathancox opened the pull request:
Fixes DOC-6743, DOC-7108, DOC-6393

Adds the new jobs PTS metrics to the docs and the `gc_expires_after` option for changefeeds. 

Updates include:

- Metrics page updated with the `job` metrics. Added `job_type` as a placeholder since there are many jobs here that fall under this metric. This breaks from the pattern of this list, however this isn't a complete list of metrics.
- Changefeed monitoring page
- Backup and restore monitoring page
- Pause job monitoring section and link to this from the Show Jobs page.
- Information around the new CDC option for expiring protected timestamp records.
- Links across docs to improve how users find information relating to jobs and metrics.

While I'm in the metrics neighborhood:

- Removed the env variable requirement for metrics labels and changed that they weren't available on Cloud products to just **not** available on Serverless. (see comment for clarification)

kathancox commented on _includes/v23.1/sidebar-data/manage.json:
```diff
@@ -161,15 +161,15 @@
+          "urls": [
+            "/${VERSION}/backup-and-restore-monitoring.html"
+          ]
+        },
```
Here I have just moved this page in the nav, so it doesn't imply that monitoring is only available for customer-owned backups

kathancox commented on v23.1/metrics.md:
```diff
@@ -5,7 +5,7 @@ toc: false
 ---
 
-As part of normal operation, CockroachDB continuously records metrics that track performance, latency, usage, and many other runtime indicators. These metrics are often useful in diagnosing problems, troubleshooting performance, or planning cluster infrastructure modifications. This page documents locations where metrics are exposed for analysis, and includes the full list of available metrics in ...
+As part of normal operation, CockroachDB continuously records metrics that track performance, latency, usage, and many other runtime indicators. These metrics are often useful in diagnosing problems, troubleshooting performance, or planning cluster infrastructure modifications. This page documents locations where metrics are exposed for analysis.
```
This page does not include the full list, so removed this.

kathancox commented on v23.1/backup-and-restore-monitoring.md:
```diff
@@ -32,29 +32,37 @@ See the [Monitor CockroachDB with Prometheus](monitor-cockroachdb-with-prometheu
 
-- Use the `schedules_backup_last_completed_time` metric to monitor the specific backup job or jobs you would use to recover from a disaster.
-- Configure alerting on the `schedules_backup_last_completed_time` metric to watch for cases where the timestamp has not moved forward as expected.
+- Use the `schedules.BACKUP.last_completed_time` metric to monitor the specific backup job or jobs you would use to recover from a disaster.
```
@miretskiy Scheduled changefeeds? Yes, we do document those... but in terms of metrics, I could not find schedules.changefeed.{metric} listed at `_status/vars`. Have they been added? Perhaps, I somehow missed those.

kathancox commented on v23.1/pause-job.md:
```diff
@@ -42,6 +42,27 @@ Parameter | Description
 
+## Monitoring paused jobs
+
+Pausing a job will release existing [protected timestamp records](architecture/storage-layer.html#protected-timestamps) that are protecting historical data from [garbage collection](architecture/storage-layer.html#garbage-collection). As a result, you could lose historical data that certain jobs need, if the job remains paused longer than the [garbage collection window](configure-replication-zone ...
```
Ack OK, I will work on updating this! @dt could you comment on the backup part here?

kathancox replied:
I adjusted this slightly to "may", but there might be more edits needed here.

amruss commented on v23.1/monitor-and-debug-changefeeds.md:
```diff
@@ -59,6 +59,17 @@ By default, changefeeds will retry errors with [some exceptions](#changefeed-ret
+
+- `jobs.changefeed.currently_paused`: Tracks the number of changefeed jobs currently considered [paused](pause-job.html). Since paused changefeed jobs can accumulate garbage, it is important to monitor the number of changefeeds paused.
+- `jobs.changefeed.expired_pts_records`: Tracks the number of expired protected timestamp records owned by changefeed jobs. You can monitor this metric in conjunction with the [`gc_protect_expires_after` option](create-changefeed.html#gc-protect-expire).
+- `jobs.changefeed.protected_age_sec`: Tracks the age of the oldest protected timestamp record protected by changefeed jobs.
```
Lets be more prescriptive here. This one is by far the most important in my opinion. I think we should move this one to the top and recommend monitoring for if protected_age_sec > the gc.ttl you have set.

kathancox replied:
Done!

amruss reviewed (approved):
Left one comment but besides that LGTM

adityamaru reviewed (changes requested):
I've reviewed the backup and general jobs changes. I'll leave the cdc parts to Yevgeniy. Should we link to https://www.cockroachlabs.com/blog/protected-timestamps-for-less-garbage/ from https://www.cockroachlabs.com/docs/v22.2/architecture/storage-layer#protected-timestamps for folks who are interested in diving further into what PTS is?

miretskiy commented on v23.1/pause-job.md:
```diff
@@ -42,6 +42,27 @@ Parameter | Description
+
+Pausing a job may release existing [protected timestamp records](architecture/storage-layer.html#protected-timestamps) that are protecting historical data from [garbage collection](architecture/storage-layer.html#garbage-collection). As a result, you could lose historical data that certain jobs need, if the job remains paused longer than the [garbage collection window](configure-replication-zones ...
+
+When a [changefeed](create-changefeed.html) job is paused, the changefeed will protect data from garbage collection up to the time of the changefeed [checkpoint](change-data-capture-overview.html#how-does-an-enterprise-changefeed-work). Changefeeds also provide options for protecting changefeed data while paused. See [Garbage collection and changefeeds](changefeed-messages.html#garbage-collection ...
```
That's correct, @adityamaru ... The changefeed *may* protect data;

kathancox replied:
I have moved to using Aditya's suggested paragraphs/format above, so this sentence has been reworked/removed.

kathancox commented on v23.1/changefeed-messages.md:
```diff
@@ -180,7 +180,17 @@ Protected timestamps will protect changefeed data from garbage collection in the
 - You [pause](pause-job.html) a changefeed with the [`protect_data_from_gc_on_pause`](create-changefeed.html#protect-pause) option enabled. Protected timestamps will protect changes until you [resume](resume-job.html) the changefeed.
 
-However, if the changefeed lags too far behind, the protected changes could cause data storage issues. To release the protected timestamps and allow garbage collection to resume, you can cancel the changefeed or [resume](resume-job.html) in the case of a paused changefeed. 
+However, if the changefeed lags too far behind, the protected changes could cause data storage issues. To release the protected timestamps and allow garbage collection to resume, you can:
```
Yes, sounds great, I have added this here.

kathancox commented on v23.1/pause-job.md:
```diff
@@ -42,6 +42,27 @@ Parameter | Description
 
+## Monitoring paused jobs
+
+Pausing a job may release existing [protected timestamp records](architecture/storage-layer.html#protected-timestamps) that are protecting historical data from [garbage collection](architecture/storage-layer.html#garbage-collection). As a result, you could lose historical data that certain jobs need, if the job remains paused longer than the [garbage collection window](configure-replication-zones ...
```
I like! I have updated the paragraphs around your suggestions here.

kathancox commented:
@adityamaru I incorporated your feedback here, would you mind taking another quick peek at the changes related to your feedback?

adityamaru reviewed (approved):
backup and jobs changes lgtm!

jayshrivastava reviewed (approved):
Changefeed changes LGTM!

rmloveland commented on v23.1/changefeed-messages.md:
```diff
@@ -180,7 +180,17 @@ Protected timestamps will protect changefeed data from garbage collection in the
+- [Resume](resume-job.html) a paused changefeed job.
+- {% include_cached new-in.html version="v23.1" %} Set the [`gc_protect_expires_after`](create-changefeed.html#gc-protect-expire) option, which will automatically expire the protected timestamp records that are older than your defined duration and cancel the changefeed job.
+
+    For example, if the following changefeed is paused or runs into an error and then pauses, protected timestamps will protect changes for up to 24 hours. After this point, if the changefeed does not resume, the protected timestamp records will expire and the changefeed job will be cancelled. This releases the protected timestamp records and allows garbage collection to resume:
```
suggest changing the text to say "... will protect changes for up to the specified interval" - that way we only have to maintain the SQL statement, not the text/statement in sync.  Also since I think this is a SQL interval, the text "interval" could be a link to https://www.cockroachlabs.com/docs/dev/interval.html

rmloveland commented on v23.1/create-changefeed.md:
```diff
@@ -157,14 +157,15 @@ Option | Value | Description
 `envelope` | `key_only` / `row`* / `wrapped` | `key_only` emits only the key and no value, which is faster if you only want to know when the key changes.<br><br>`row` emits the row without any additional metadata fields in the message. *You can only use `row` with Kafka sinks or sinkless changefeeds. `row` does not support [`avro` format](#format).<br><br>`wrapped` emits the full message includin ...
 <a name="format"></a>`format` | `json` / `avro` / `csv`* | Format of the emitted record. For mappings of CockroachDB types to Avro types, [see the table](changefeed-messages.html#avro-types) and detail on [Avro limitations](changefeed-messages.html#avro-limitations). <br><br>*`format=csv` works only in combination with [`initial_scan = 'only'`](#initial-scan). You cannot combine `format=csv` with ...
 <a name="full-table-option"></a>`full_table_name` | N/A | Use fully qualified table name in topics, subjects, schemas, and record output instead of the default table name. This can prevent unintended behavior when the same table name is present in multiple databases.<br><br>**Note:** This option cannot modify existing table names used as topics, subjects, etc., as part of an [`ALTER CHANGEFEED`]( ...
+<span class="version-tag">New in v23.1:</span> <a name="gc-protect-expire"></a>`gc_protect_expires_after` | [Duration string](https://pkg.go.dev/time#ParseDuration) | Automatically expires protected timestamp records that are older than the defined duration. In the case where a changefeed job remains paused, `gc_protect_expires_after` will trigger the underlying protected timestamp record to expi ...
```
oh ok its a Golang duration.  As I mentioned `INTERVAL` in my previous comment, is this also a SQL `INTERVAL` ?  would be nice if so.  

If not then I guess amend my previous comment to say instead of using the text "interval" and linking to `INTERVAL` docs maybe just edit the text referred to in my previous comment to say "duration" with no link

suggest making reference to "protected timestamp" some kind of link to protected timestamps docs.  Esp. since this is an anchor link so folks may jump here directly and not get surrounding context

similarly suggest making the following text links: "change feed job".  Suggest changing "accumulation of protected data" to "accumulation of protected MVCC values" and make MVCC a link to https://www.cockroachlabs.com/docs/v23.1/architecture/storage-layer.html#mvcc

rmloveland replied:
oh ok its a Golang duration.  As I mentioned `INTERVAL` in my previous comment, is this also a SQL `INTERVAL` ?  would be nice if so.  

If not then I guess amend my previous comment to say instead of using the text "interval" and linking to `INTERVAL` docs maybe just edit the text referred to in my previous comment to say "duration" with no link

suggest making reference to "protected timestamp" some kind of link to protected timestamps docs.  Esp. since this is an anchor link so folks may jump here directly and not get surrounding context

similarly suggest making the following text links: "change feed job".  Suggest changing "accumulation of protected data" to "accumulation of protected MVCC values" and make MVCC a link to https://www.cockroachlabs.com/docs/v23.1/architecture/storage-layer.html#mvcc

rmloveland commented on v23.1/create-changefeed.md:
```diff
@@ -157,14 +157,15 @@ Option | Value | Description
 `mvcc_timestamp` | N/A |  Include the [MVCC](architecture/storage-layer.html#mvcc) timestamp for each emitted row in a changefeed. With the `mvcc_timestamp` option, each emitted row will always contain its MVCC timestamp, even during the changefeed's initial backfill.
 <a name="on-error"></a>`on_error` | `pause` / `fail` |  Use `on_error=pause` to pause the changefeed when encountering **non**-retryable errors. `on_error=pause` will pause the changefeed instead of sending it into a terminal failure state. **Note:** Retryable errors will continue to be retried with this option specified. <br><br>Use with [`protect_data_from_gc_on_pause`](#protect-pause) to prote ...
-<a name="protect-pause"></a>`protect_data_from_gc_on_pause` | N/A |  When a [changefeed is paused](pause-job.html), ensure that the data needed to [resume the changefeed](resume-job.html) is not garbage collected. If `protect_data_from_gc_on_pause` is **unset**, pausing the changefeed will release the existing protected timestamp records. It is also important to note that pausing and adding `prot ...
+<a name="protect-pause"></a>`protect_data_from_gc_on_pause` | N/A |  When a [changefeed is paused](pause-job.html), ensure that the data needed to [resume the changefeed](resume-job.html) is not garbage collected. If `protect_data_from_gc_on_pause` is **unset**, pausing the changefeed will release the existing protected timestamp records. It is also important to note that pausing and adding `prot ...
```
suggest making "garbage collected" a link to the storage layer GC docs you referenced elsewhere

suggest "protected timestamp" be a link to docs on that concept 

same comment as elsewhere to have "non-retryable" be a link to some docs on that, if we do have any

suggest "can prevent garbage collection" be extended to add something like " ... which will negatively impact write performance, including DELETEs" or something like that, so folks know it's not without a cost, could save a support issue maybe

rmloveland commented on v23.1/manage-a-backup-schedule.md:
```diff
@@ -40,17 +40,19 @@ Further guidance on connecting to Amazon S3, Google Cloud Storage, Azure Storage
 ## Set up monitoring for the backup schedule
 
-We recommend that you [monitor your backup schedule with Prometheus](monitoring-and-alerting.html#prometheus-endpoint), and alert when there are anomalies such as backups that have failed or no backups succeeding over a certain amount of time&mdash; at which point, you can inspect schedules by running [`SHOW SCHEDULES`](show-schedules.html).
+We recommend that you [monitor your backup schedule with Prometheus](monitoring-and-alerting.html#prometheus-endpoint), and alert when there are anomalies such as backups that have failed or no backups succeeding over a certain amount of time—at which point, you can inspect schedules by running [`SHOW SCHEDULES`](show-schedules.html).
```
I'm sorry as an ascii heathen I do not understand what is this character 

(jkjk)

rmloveland commented on v23.1/manage-a-backup-schedule.md:
```diff
@@ -40,17 +40,19 @@ Further guidance on connecting to Amazon S3, Google Cloud Storage, Azure Storage
 
-        When `schedules_BACKUP_failed` increments, run [`SHOW SCHEDULES`](show-schedules.html) to check which schedule is affected and to inspect the error in the `status` column.
+        When `schedules.BACKUP.failed` increments, run [`SHOW SCHEDULES`](show-schedules.html) to check which schedule is affected and to inspect the error in the `status` column.
+    - {% include_cached new-in.html version="v23.1" %} `schedules.BACKUP.protected_age_sec`: The age of the oldest protected timestamp record protected by backup schedules.
```
suggest adding links to protected timestamps docs

rmloveland commented on v23.1/monitor-and-debug-changefeeds.md:
```diff
@@ -59,6 +59,17 @@ By default, changefeeds will retry errors with [some exceptions](#changefeed-ret
 
+#### Protected timestamp and garbage collection monitoring
+
+Protected timestamps will protect changefeed data from garbage collection in particular scenarios, but if a changefeed lags too far behind, the protected changes could cause data storage issues. See [Garbage collection and changefeeds](changefeed-messages.html#garbage-collection-and-changefeeds) for detail on when changefeed data is protected from garbage collection.
```
suggest first mention of protected timestamps be a link to the pts stuff in architecture docs

I think "could cause data storage issues" would be more impactful if more concrete, something like "the protected changes will cause an accumulation of [MVCC values](https://www.cockroachlabs.com/docs/v23.1/architecture/storage-layer.html#mvcc) which in addition to increasing disk usage will negatively impact writes to those values, including DELETEs" ... or similar?  you would probably want an engineer to vet that text so understand if you need to punt for now given release deadline

rmloveland commented on v23.1/show-jobs.md:
```diff
@@ -98,6 +98,10 @@ Status | Description
 `retrying` | Job is retrying another job that failed.
 
+{{site.data.alerts.callout_info}}
+We recommend monitoring paused jobs to protect historical data from [garbage collection](architecture/storage-layer.html#garbage-collection), or potential data accumulation in the case of [changefeeds](changefeed-messages.html#garbage-collection-and-changefeeds). See [Monitoring paused jobs](pause-job.html#monitoring-paused-jobs) for detail on metrics to track paused jobs and [protected timestamp ...
```
does "data accumulation" here and elsewhere mean "MVCC value accumulation" ?  if so I recommend the latter phrase, and making MVCC a link to https://www.cockroachlabs.com/docs/v23.1/architecture/storage-layer.html#mvcc

rmloveland reviewed (approved):
LGTM, most of my comments are about adding links and other non-blocking stuff. 

In a few cases I've asked if we could be a bit more specific around "Data accumulation" meaning "MVCC value accumulation", and whether "degraded performance for some workloads" could be more specific w.r.t. the some workloads being "write-heavy workloads including DELETEs" with a link to https://www.cockroachlabs.com/docs/dev/delete.html#preserving-delete-performance-over-time which I think explains the issue nicely.  But as I say below these updates probably require final confirmation from an engineer and could be elided for now given the release timing

kathancox commented:
> LGTM, most of my comments are about adding links and other non-blocking stuff.
> 
> In a few cases I've asked if we could be a bit more specific around "Data accumulation" meaning "MVCC value accumulation", and whether "degraded performance for some workloads" could be more specific w.r.t. the some workloads being "write-heavy workloads including DELETEs" with a link to https://www.cockroachlabs.com/docs/dev/delete.html#preserving-delete-performance-over-time which I think explains the issue nicely. But as I say below these updates probably require final confirmation from an engineer and could be elided for now given the release timing

Thanks @rmloveland! I'm going to add all the links here.

I have also added to the following issue on the metrics-names.md file (https://cockroachlabs.atlassian.net/browse/DOC-4879). This page was difficult to work with. I added one link to the table in my original work, but had kept it to a minimum since the rest of the table seemed to not do that. However, since I've broken that consistency once, I might as well provide as much information on these metrics for now. This page needs a bit of work and I've noted that there in the issue.

I've created this issue: https://cockroachlabs.atlassian.net/browse/DOC-7620 to give a bit more detail for users on "data accumulation". Thanks for the suggestion here. As you pointed to, I'm going to work on that after this PR is merged so that the main 23.1 features here are merged for release.

As always, appreciate you keeping me honest with the links! 😸

rmloveland commented:
Kathryn Hancox ***@***.***> writes:

>>     LGTM, most of my comments are about adding links and other
>>     non-blocking stuff.
>>
>>     In a few cases I've asked if we could be a bit more specific
>>     around "Data accumulation" meaning "MVCC value accumulation", and
>>     whether "degraded performance for some workloads" could be more
>>     specific w.r.t. the some workloads being "write-heavy workloads
>>     including DELETEs" with a link to
>>     https://www.cockroachlabs.com/docs/dev/delete.html#preserving-delete-performance-over-time
>>     which I think explains the issue nicely. But as I say below these
>>     updates probably require final confirmation from an engineer and
>>     could be elided for now given the release timing

> Thanks @rmloveland! I'm going to add all the links here.

Sweet thanks!

> I have also added to the following issue on the metrics-names.md file
> (https://cockroachlabs.atlassian.net/browse/DOC-4879). This page was
> difficult to work with. I added one link to the table in my original
> work, but had kept it to a minimum since the rest of the table seemed
> to not do that. However, since I've broken that consistency once, I
> might as well provide as much information on these metrics for now.
> This page needs a bit of work and I've noted that there in the issue.

> I've created this issue:
> https://cockroachlabs.atlassian.net/browse/DOC-7620 to give a bit more
> detail for users on "data accumulation". Thanks for the suggestion
> here. As you pointed to, I'm going to work on that after this PR is
> merged so that the main 23.1 features here are merged for release.

Cool, ya I def think most of what I was talking about is "nice to have"
and will hold til post release

> As always, appreciate you keeping me honest with the links! 😸

lol

Sincerely, the "Where's the Links" guy
